### PR TITLE
Remove fastify-websocket, proxy ws events directly

### PR DESCRIPTION
**Pull request by greguz**
## Notable changes:
- Remove `fastify-websocket`: proxy `ws` events directly.
- Proxy query string with WebSockets.
- Support `prefix` and `rewritePrefix` options with WebSockets.
- Support custom `ws` options for both client and server.
- Add `socket.io` tests.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,10 +1,90 @@
...
+
+  return target
+}
+
+function setupWebSocketProxy (fastify, options) {
+  const server = new WebSocket.Server({
+    path: options.prefix,
+    server: fastify.server,
+    ...options.wsServerOptions
+  })
```
Comment (mcollina) [inline index.js:74]: This logic is missing: https://github.com/fastify/fastify-websocket/blob/99c94a003da2353c1f482301027166e64e743582/index.js#L71-L83

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work!

File: index.js (hunk 1 above)
Comment (greguz) [reply index.js:74]: Just added again using the `onClose` hook. It seems that this hook was not available on older Fastify versions, so I tried to use `fastify-plugin` for version checking, but this breaks the `prefix` option for HTTP routes. Can we just assume that, in 2020, Fastify must have the `onClose` hook? :P

Comment (mcollina) [reply index.js:74]: Yes. You can also set https://github.com/fastify/fastify-plugin/blob/52cec31e4c0dc4eb6ecc2b2f0e2f1c21994713d1/plugin.js#L49 as the actual check is in Fastify itself.

File: index.js [hunk 2]
```diff
@@ -1,10 +1,97 @@
...
+      client.close()
+    }
+    done()
+  })
+
+  server.on('connection', (source, request) => {
+    const url = createWebSocketUrl(options, request)
+
+    const target = new WebSocket(url, {
+      rejectUnauthorized: process.env.NODE_ENV === 'production',
```
Comment (mcollina) [inline index.js:87]: We should not be doing this, it should be an option.

File: index.js [hunk 3]
```diff
@@ -1,10 +1,97 @@
...
+    server: fastify.server,
+    ...options.wsServerOptions
+  })
+
+  fastify.addHook('onClose', (instance, done) => {
+    for (const client of server.clients) {
+      client.close()
+    }
+    done()
+  })
```
Comment (mcollina) [inline index.js:81]: now that this is responsibility of this plugin, we need a test for this logic.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (greguz) [reply index.js]: Done.

File: index.js (hunk 3 above)
Comment (greguz) [reply index.js]: Added close event await to WebSocket tests.

File: index.js (hunk 1 above)
Comment (greguz) [reply index.js:74]: I missed the WebSocket server close, and while writing tests, I actually understood the motivation behind that monkey-path.

File: test/websocket.js [hunk 4]
```diff
@@ -33,7 +33,6 @@ test('basic websocket proxy', async (t) => {
   })
 
   await server.listen(0)
-  t.tearDown(server.close.bind(server))
```
Comment (mcollina) [inline test/websocket.js:36]: This really need to be there.

File: test/socket.io.js [hunk 5]
```diff
@@ -21,7 +21,6 @@ test('proxy socket.io', async t => {
   await promisify(srvUpstream.listen.bind(srvUpstream))(0)
 
   const srvProxy = Fastify()
-  t.tearDown(srvProxy.close.bind(srvProxy))
```
Comment (mcollina) [inline test/socket.io.js]: This really need to be there.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/websocket.js (hunk 4 above)
Comment (greguz) [reply test/websocket.js]: Done.

File: test/socket.io.js (hunk 5 above)
Comment (greguz) [reply test/socket.io.js]: Done.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**